core/include/cond.h

```diff
@@ -0,0 +1,223 @@
+ * is the standard way to use Mesa-style condition variables.
+ *
+ * Example: Suppose we want to implement a bounded queue, such as a Unix-style
+ * pipe between two threads. When data is written to the pipe, it is appended
+ * to a queue, and the writing thread blocks if the queue is full. When data
+ * is read from the pipe, it is removed from the queue; if the queue is empty,
+ * the reading thread blocks until it is not empty. If the pipe is closed by
+ * the sender, waiting reading threads wake up.
+ *
+ * Here is a sketch of how to implement such a structure with condition
+ * variables. For simplicity, messages are single bytes. We assume a FIFO data
+ * structure queue_t. We assume it is unsafe to add to the queue if it is full,
+ * or remove from the queue if it is empty.
+ *
+ * typedef struct pipe {
+ *     queue_t queue;
+ *     cond_t read_cond;
+ *     cond_t write_cond;
+ *     mutex_t lock;
+ *     bool closed;
+ * } pipe_t;
+ *
+ * void pipe_init(pipe_t* pipe) {
+ *     queue_init(&pipe->queue);
+ *     cond_init(&pipe->read_cond);
+ *     cond_init(&pipe->write_cond);
+ *     mutex_init(&pipe->lock);
+ *     pipe->closed = false;
+ * }
+ *
+ * void pipe_write(pipe_t* pipe, char c) {
+ *     mutex_lock(&pipe->lock);
+ *     while (queue_length(&pipe->queue) == MAX_QUEUE_LENGTH && !pipe->closed) {
+ *         cond_wait(&pipe->write_cond, &pipe->lock);
+ *     }
+ *     if (pipe->closed) {
+ *         mutex_unlock(&pipe->lock);
+ *         return 0;
+ *     }
+ *     add_to_queue(&pipe->queue, c);
+ *     cond_signal(&pipe->read_cond);
+ *     mutex_unlock(&pipe->lock);
+ *     return 1;
+ * }
+ *
+ * void pipe_close(pipe_t* pipe) {
+ *     mutex_lock(&pipe->lock);
+ *     pipe->closed = true;
+ *     cond_broadcast(&pipe->read_cond);
+ *     cond_broadcast(&pipe->write_cond);
+ *     mutex_unlock(&pipe->lock);
+ * }
+ *
+ * int pipe_read(pipe_t* pipe, char* buf) {
+ *     mutex_lock(&pipe->lock);
+ *     while (queue_length(&pipe->queue) == 0 && !pipe->closed) {
+ *         cond_wait(&pipe->read_cond, &pipe->lock);
+ *     }
+ *     if (pipe->closed) {
+ *         mutex_unlock(&pipe->lock);
+ *         return 0;
+ *     }
+ *     *buf = remove_from_queue(&pipe->queue);
+ *     cond_signal(&pipe->write_cond);
+ *     mutex_unlock(&pipe->lock);
+ *     return 1;
+ * }
+ *
+ * Note that this could actually be written with a single condition variable.
+ * However, the example includes two for didactic reasons.
+ *
+ * @author      Sam Kumar <samkumar@berkeley.edu>
+ */
+
+#ifndef COND_H
+#define COND_H
+
+#include <stdbool.h>
+#include <stddef.h>
+
+#include "list.h"
+#include "mutex.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @brief Condition variable structure. Must never be modified by the user.
+ */
+typedef struct {
+    /**
+     * @brief   The process waiting queue of the condition variable.
+     * @internal
+     */
+    list_node_t queue;
+} cond_t;
+
+/**
+ * @brief   Static initializer for cond_t.
```

Comment:
add empty line after `@brief`

Reply:
I've added an empty line after `@brief` in all comments.